release/releaser.go

```diff
@@ -45,539 +36,318 @@ func NewReleaser(
-		Description: fmt.Sprintf("Release %d service(s): %s.", len(services), strings.Join(service2string(services), ", ")),
-		Do: func(rc *ReleaseContext) (res string, err error) {
-			cause := strconv.Quote(msg)
-
-			// We'll collect results for each service release.
-			results := map[flux.ServiceID]error{}
-
-			// Collect definitions for each service release.
-			var defs []platform.ServiceDefinition
-			// If we're regrading our own image, we want to do that
-			// last, and "asynchronously" (meaning we probably won't
-			// see the reply).
-			var asyncDefs []platform.ServiceDefinition
-
-			for _, service := range services {
-				def, ok := rc.PodControllers[service]
-				if !ok {
-					results[service] = errors.New("no definition found; skipping release")
-					continue
-				}
-
-				namespace, serviceName := service.Components()
-				switch serviceName {
-				case FluxServiceName, FluxDaemonName:
-					rc.Instance.LogEvent(namespace, serviceName, "Starting "+cause+". (no result expected)")
-					asyncDefs = append(asyncDefs, platform.ServiceDefinition{
-						ServiceID:     service,
-						NewDefinition: def,
-						Async:         true,
-					})
-				default:
-					rc.Instance.LogEvent(namespace, serviceName, "Starting "+cause)
-					defs = append(defs, platform.ServiceDefinition{
-						ServiceID:     service,
-						NewDefinition: def,
-					})
-				}
-			}
-
-			// Execute the releases as a single transaction.
-			// Splat any errors into our results map.
-			transactionErr := rc.Instance.PlatformApply(defs)
-			if transactionErr != nil {
-				switch err := transactionErr.(type) {
-				case platform.ApplyError:
-					for id, applyErr := range err {
-						results[id] = applyErr
-					}
-				default: // assume everything failed, if there was a coverall error
-					for _, service := range services {
-						results[service] = transactionErr
-					}
-				}
-			}
-
-			// Report individual service release results.
-			for _, service := range services {
-				namespace, serviceName := service.Components()
-				switch serviceName {
-				case FluxServiceName, FluxDaemonName:
-					continue
-				default:
-					if err := results[service]; err == nil { // no entry = nil error
-						rc.Instance.LogEvent(namespace, serviceName, msg+". done")
-					} else {
-						rc.Instance.LogEvent(namespace, serviceName, msg+". error: "+err.Error()+". failed")
-					}
-				}
-			}
-
-			// Lastly, services for which we don't expect a result
-			// (i.e., ourselves). This will kick off the release in
-			// the daemon, which will cause Kubernetes to restart the
-			// service. In the meantime, however, we will have
-			// finished recording what happened, as part of a graceful
-			// shutdown. So the only thing that goes missing is the
-			// result from this release call.
-			if len(asyncDefs) > 0 {
-				rc.Instance.PlatformApply(asyncDefs)
-			}
-
-			return "", transactionErr
-		},
+func commitMessageFromReleaseSpec(spec *flux.ReleaseSpec) string {
```

Comment:
Ah, though maybe the notification should be based on what it *actually* did, not just the intent. Since the actual results are obvious from the git commit, this can just be the intent.